File: android/src/main/kotlin/jyahann/dgis_map_kit/MarkersUtils.kt
package jyahann.dgis_map_kit

import ru.dgis.sdk.coordinates.GeoPoint
import ru.dgis.sdk.geometry.Elevation
import ru.dgis.sdk.geometry.GeoPointWithElevation
import ru.dgis.sdk.map.Anchor
import ru.dgis.sdk.map.Color
import ru.dgis.sdk.map.LogicalPixel
import ru.dgis.sdk.map.MapDirection
import ru.dgis.sdk.map.Marker
import ru.dgis.sdk.map.MarkerOptions
import ru.dgis.sdk.map.Opacity
import ru.dgis.sdk.map.SimpleClusterOptions
import ru.dgis.sdk.map.TextPlacement
import ru.dgis.sdk.map.TextStyle
import ru.dgis.sdk.map.ZIndex
import ru.dgis.sdk.map.imageFromAsset

class MarkersUtils {
    companion object {
        fun getMarkerFromDart(marker: Map<String, Any?>, sdkContext: ru.dgis.sdk.Context, layerId: String?): Marker {
            val iconOptions = marker["iconOptions"] as Map<String, Any>
            var assetLookupKey =
                io.flutter.view.FlutterMain.getLookupKeyForAsset(marker["icon"] as String)

            return Marker(
                MarkerOptions(
                    position = GeoPointWithElevation(
                            CameraUtils.getGeoPointFromDart(
                                marker["position"] as Map<String, Any>
                            ),
                            elevation = Elevation((marker["elevation"] as Number).toFloat())
                        ),
                    icon = imageFromAsset(sdkContext, assetLookupKey),
                    anchor = getAnchorFromDart(iconOptions["anchor"] as Map<String, Any>),
                    text = iconOptions["text"] as String?,
                    zIndex = ZIndex((iconOptions["zIndex"] as Number).toInt()),
                    userData = MapObjectUserData(
                        type = MapObjectUserDataType.MARKER,
                        userData = marker,
                        layerId = layerId,
                    ),
                    iconMapDirection = getMapDirectionFromDart(iconOptions["iconMapDirection"]),
                    iconOpacity = Opacity((iconOptions["iconOpacity"] as Number).toFloat()),
                    animatedAppearance = iconOptions["animatedAppearance"] as Boolean,
                    iconWidth = LogicalPixel((iconOptions["size"] as Number).toFloat()),
                    textStyle =
                    getTextStyleFromDart(
                        iconOptions["textStyle"] as Map<String, Any>,
                    )
                )
            )
        }

        fun getClusterOptionsFromDart(
            mapClusterer: Map<String, Any?>,
            markers: List<Any?>,
            sdkContext: ru.dgis.sdk.Context,
            layerId: String?
        ): SimpleClusterOptions {
            val iconOptions = mapClusterer["iconOptions"] as Map<String, Any>
            var assetLookupKey =
                io.flutter.view.FlutterMain.getLookupKeyForAsset(mapClusterer["icon"] as String)

            return SimpleClusterOptions(
                icon = imageFromAsset(sdkContext, assetLookupKey),
                anchor = getAnchorFromDart(iconOptions["anchor"] as Map<String, Any>),
                text = iconOptions["text"] as String?,
                zIndex = ZIndex((iconOptions["zIndex"] as Number).toInt()),
                userData =
                MapObjectUserData(
                    type = MapObjectUserDataType.CLUSTER,
                    userData = markers,
                    layerId = layerId,
                ),
                iconMapDirection = getMapDirectionFromDart(iconOptions["iconMapDirection"]),
                iconOpacity = Opacity((iconOptions["iconOpacity"] as Number).toFloat()),
                animatedAppearance = iconOptions["animatedAppearance"] as Boolean,
                textStyle =
                getTextStyleFromDart(
                    iconOptions["textStyle"] as Map<String, Any>,
                ),
                iconWidth = LogicalPixel((iconOptions["size"] as Number).toFloat())
            )
        }

        fun getMapDirectionFromDart(value: Any?): MapDirection? {
            if (value == null) {
                return null;
            }
            return MapDirection(value = (value as Number).toDouble())
        }

        fun getAnchorFromDart(anchor: Map<String, Any>): Anchor {
            return Anchor(x = (anchor["x"] as Number).toFloat(), y = (anchor["y"] as Number).toFloat())
        }

        fun getTextStyleFromDart(textStyle: Map<String, Any>): TextStyle {
            return TextStyle(
                fontSize = LogicalPixel((textStyle["fontSize"] as Number).toFloat()),
                color = Color((textStyle["color"] as Number).toInt()),
                strokeWidth = LogicalPixel((textStyle["strokeWidth"] as Number).toFloat()),
                strokeColor = Color((textStyle["strokeColor"] as Number).toInt()),
                textPlacement = TextPlacement.valueOf(textStyle["textPlacement"] as String),
                textOffset = LogicalPixel((textStyle["textOffset"] as Number).toFloat(),),
            )
        }
    }
}

File: ios/Classes/MarkersUtils.swift
import DGis
import Flutter

class MarkersUtils {
    static func getClusterOptionsFromDart(
        mapClusterer: Dictionary<String, Any?>,
        sdk: DGis.Container,
        registrar: FlutterPluginRegistrar,
        markers: Array<Any>,
        layerId: String?
    ) -> SimpleClusterOptions {
        let iconOptions = mapClusterer["iconOptions"] as! Dictionary<String, Any>;
        let assetLookupKey = registrar.lookupKey(forAsset: mapClusterer["icon"] as! String);
        let image = UIImage.loadFromFile(imagePath: assetLookupKey);
        let icon = sdk.imageFactory.make(image: image!)
        
        return SimpleClusterOptions(
            icon: icon,
            iconMapDirection: getMapDirectionFromDart(
                value: iconOptions["iconMapDirection"]
            ),
            anchor: getAnchorFromDart(
                anchor: iconOptions["anchor"] as! Dictionary<String, Any>
            ),
            text: iconOptions["text"] as? String,
            textStyle: getTextStyleFromDart(
                textStyle: iconOptions["textStyle"] as! Dictionary<String, Any>
            ),
            iconOpacity: Opacity(
                value: Float(iconOptions["iconOpacity"] as! Double)
            ),            iconWidth: LogicalPixel(
                value: Float(iconOptions["size"] as! Double)
            ),
            userData: MapObjectUserData(
                userDataType: MapObjectUserDataType.cluster,
                userData: markers,
                layerId: layerId
            ),
            zIndex: ZIndex(
                value: UInt32(iconOptions["zIndex"] as! Double)
            ),
            animatedAppearance: iconOptions["animatedAppearance"] as! Bool
        );
    }
    
    static func getLayerOptionsFromDart(
        marker: Dictionary<String, Any?>,
        sdk: DGis.Container,
        registrar: FlutterPluginRegistrar,
        layerId: String?
    ) -> SimpleClusterOptions {
        let iconOptions = marker["iconOptions"] as! Dictionary<String, Any>;
        let assetLookupKey = registrar.lookupKey(forAsset: marker["icon"] as! String);
        let image = UIImage.loadFromFile(imagePath: assetLookupKey);
        let icon = sdk.imageFactory.make(image: image!)
        
        return SimpleClusterOptions(
            icon: icon,
            iconMapDirection: getMapDirectionFromDart(
                value: iconOptions["iconMapDirection"]
            ),
            anchor: getAnchorFromDart(
                anchor: iconOptions["anchor"] as! Dictionary<String, Any>
            ),
            text: iconOptions["text"] as? String,
            textStyle: getTextStyleFromDart(
                textStyle: iconOptions["textStyle"] as! Dictionary<String, Any>
            ),
            iconOpacity: Opacity(
                value: Float(iconOptions["iconOpacity"] as! Double)
            ),            iconWidth: LogicalPixel(
                value: Float(iconOptions["size"] as! Double)
            ),
            userData: MapObjectUserData(
                userDataType: MapObjectUserDataType.marker,
                userData: marker,
                layerId: layerId
            ),
            zIndex: ZIndex(
                value: UInt32(iconOptions["zIndex"] as! Double)
            ),
            animatedAppearance: iconOptions["animatedAppearance"] as! Bool
        );
    }
    
    static func getMarkerFromDart(
        marker: Dictionary<String, Any?>,
        sdk: DGis.Container,
        registrar: FlutterPluginRegistrar,
        layerId: String?
    ) -> Marker {
        let iconOptions = marker["iconOptions"] as! Dictionary<String, Any>;
        let assetLookupKey = registrar.lookupKey(forAsset: marker["icon"] as! String);
        let image = UIImage.loadFromFile(imagePath: assetLookupKey);
        let icon = sdk.imageFactory.make(image: image!)
        
        let options = MarkerOptions(
            position: GeoPointWithElevation(
                point: CameraUtils.getGeoPointFromDart(
                    coordinates: marker["position"] as! Dictionary<String, Any>
                ),
                elevation: Elevation(marker["elevation"] as! Double)
            ),
            icon: icon,
            iconMapDirection: getMapDirectionFromDart(
                value: iconOptions["iconMapDirection"]
            ), 
            anchor: getAnchorFromDart(
                anchor: iconOptions["anchor"] as! Dictionary<String, Any>
            ),
            text: iconOptions["text"] as? String,
            textStyle: getTextStyleFromDart(
                textStyle: iconOptions["textStyle"] as! Dictionary<String, Any>
            ),
            iconOpacity: Opacity(
                value: Float(iconOptions["iconOpacity"] as! Double)
            ),
            visible: true,
            draggable: false,
            iconWidth: LogicalPixel(
                value: Float(iconOptions["size"] as! Double)
            ),
            userData: MapObjectUserData(
                userDataType: MapObjectUserDataType.marker,
                userData: marker,
                layerId: layerId
            ),
            zIndex: ZIndex(
                value: UInt32(iconOptions["zIndex"] as! Double)
            ),
            animatedAppearance: iconOptions["animatedAppearance"] as! Bool
        );
        
        return Marker(options: options);
    }
    
    static func getTextStyleFromDart(textStyle: Dictionary<String, Any>) -> TextStyle {
        return TextStyle(
            fontSize: LogicalPixel(
                value: Float(textStyle["fontSize"] as! Double)
            ),
            color: Color(
                argb: UInt32(textStyle["color"] as! Double)
            ),
            strokeWidth: LogicalPixel(
                value: Float(textStyle["strokeWidth"] as! Double)
            ),
            strokeColor: Color(
                argb: UInt32(textStyle["strokeColor"] as! Double)
            ),
            textPlacement: TextPlacement.fromString(
                value: textStyle["textPlacement"] as! String
            ),
            textOffset: LogicalPixel(
                value: Float(textStyle["textOffset"] as! Double)
            )
        );
    }
    
    static func getMapDirectionFromDart(value: Any?) -> MapDirection? {
        if (value is NSNull) {
            return nil;
        }
        return MapDirection(
            value: value as! Double
        );
    }
    
    static func getAnchorFromDart(anchor: Dictionary<String, Any>) -> Anchor {
        return Anchor(
            x: Float(anchor["x"] as! Double),
            y: Float(anchor["y"] as! Double)
        )
    }
}
